openl3/core.py

```diff
@@ -209,15 +214,327 @@ def process_file(filepath, output_dir=None, suffix=None, model=None,
-                                  hop_size=hop_size, verbose=1 if verbose else 0)
+    embedding, ts = get_audio_embedding(audio, sr, model=model, input_repr=input_repr,
+                                        content_type=content_type,
+                                        embedding_size=embedding_size, center=center,
+                                        hop_size=hop_size, verbose=1 if verbose else 0)
 
     np.savez(output_path, embedding=embedding, timestamps=ts)
     assert os.path.exists(output_path)
 
 
+def _preprocess_image_batch(image):
+    """
+    Preprocesses an image array so that they are rescaled and cropped to the
+    appropriate dimensions required by the embedding model.
+
+    Parameters
+    ----------
+    image : np.ndarray [shape=(H, W, C) or (N, H, W, C)]
+        3D or 4D numpy array of image data. If the images are not 224x224,
+        the images are resized so that the smallest size is 256 and then
+        the center 224x224 patch is extracted from the images. Any type
+        is accepted, and will be converted to np.float32 in the range [-1,1].
+        Signed data-types are assumed to take on negative values.
+
+    Returns
+    -------
+    batch : np.ndarray [shape=(N, H, W, C)]
+        4d numpy array of image data.
+    """
+    if image.ndim == 3:
+        # Add a batch dimension dimension
+        image = image[np.newaxis, ...]
+
+    if min(image.shape[1], image.shape[2]) < 224:
+        err_msg = 'Image(s) must be at at least as large as 224x224 px. ' \
+                  'Got image(s) of size {}x{} px'
+        raise OpenL3Error(err_msg.format(image.shape[1], image.shape[2]))
+
+    if image.shape[1] != 224 or image.shape[2] != 224:
+        # If image is not 224x224, rescale to 256x256, and take center
+        # 224x224 image patch, corresponding to what was done in L3
+        scaling = 256.0 / min(image.shape[1], image.shape[2])
+        batch = np.zeros((image.shape[0], 224, 224, 3))
+        for idx, frame in enumerate(image):
+            # Only reshape if image is larger than 256x256
+            if min(frame.shape[0], frame.shape[1]) > 256:
+                frame = skimage.transform.rescale(frame, scaling)
+            x1, x2 = frame.shape[:-1]
+            startx1 = x1//2-(224//2)
+            startx2 = x2//2-(224//2)
+            batch[idx] = frame[startx1:startx1+224,startx2:startx2+224]
+    else:
+        batch = image
+
+    # Make sure image correct type
+    if batch.dtype in (np.float16, np.float32, np.float64, np.int8,
+                       np.int16, np.int32, np.int64):
+        batch = skimage.img_as_float32(batch)
+    elif batch.dtype in (np.uint8, np.uint16, np.uint32, np.uint64):
+        # If unsigned int, convert to range [-1, 1]
+        batch = 2 * skimage.img_as_float32(batch) - 1
+
+    # Make sure maximum magnitude is in the range [-1, 1]
+    if np.max(np.abs(batch)) > 1:
+        batch /= np.max(np.abs(batch))
```

Comment:
this normalizes the entire batch based on the max value in the batch. Is that really what we want to do? Don't we want to normalize each image independently of the others based on its own values?